extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/IncrementalPublishingKafkaIndexTaskRunner.java

```diff
@@ -126,38 +126,56 @@ private void possiblyResetOffsetsOrWait(
       TaskToolbox taskToolbox
   ) throws InterruptedException, IOException
   {
-    final Map<TopicPartition, Long> resetPartitions = new HashMap<>();
-    boolean doReset = false;
+    final Map<TopicPartition, Long> newOffsetInMetadata = new HashMap<>();
+
     if (task.getTuningConfig().isResetOffsetAutomatically()) {
       for (Map.Entry<TopicPartition, Long> outOfRangePartition : outOfRangePartitions.entrySet()) {
         final TopicPartition topicPartition = outOfRangePartition.getKey();
-        final long nextOffset = outOfRangePartition.getValue();
-        // seek to the beginning to get the least available offset
+        final long outOfRangeOffset = outOfRangePartition.getValue();
+
         StreamPartition<Integer> streamPartition = StreamPartition.of(
             topicPartition.topic(),
             topicPartition.partition()
         );
-        final Long leastAvailableOffset = recordSupplier.getEarliestSequenceNumber(streamPartition);
-        if (leastAvailableOffset == null) {
-          throw new ISE(
-              "got null sequence number for partition[%s] when fetching from kafka!",
-              topicPartition.partition()
-          );
+
+        final Long earliestAvailableOffset = recordSupplier.getEarliestSequenceNumber(streamPartition);
+        if (earliestAvailableOffset == null) {
+          throw new ISE("got null earliest sequence number for partition[%s] when fetching from kafka!",
+                        topicPartition.partition());
         }
-        // reset the seek
-        recordSupplier.seek(streamPartition, nextOffset);
-        // Reset consumer offset if resetOffsetAutomatically is set to true
-        // and the current message offset in the kafka partition is more than the
-        // next message offset that we are trying to fetch
-        if (leastAvailableOffset > nextOffset) {
-          doReset = true;
-          resetPartitions.put(topicPartition, nextOffset);
+
+        if (outOfRangeOffset < earliestAvailableOffset) {
```

Comment:
I think this logic can further be refined here to call reset within this block itself. Also, it looks like an `OffsetOutOfRangeException` is thrown when the offset for the partition is either larger or smaller than the range of offsets the server has for the given partition. So the case of earliestAvailableOffset <= outofRangeOffset <= latestAvailableOffset doesn't apply. 

So it could look something like this:
```
        if (outOfRangeOffset < earliestAvailableOffset) {
          // In this case, it's probably because the messages are no longer in the Kafka cluster i.e. the messages in
          // [outOfRangeOffset, earliestAvailableOffset) are lost. Since these lost messages can no longer be
          // recovered, it's reasonable to reset the offset to the earliest available position to help ingestion resume.

          logger.warn("Seeking kafka offset to earliest offset: " + earliestAvailableOffset);
          recordSupplier.seek(streamPartition, earliestAvailableOffset);
          // TBD: still need to confirm if this should be outOfRangeOffset or earliestAvailableOffset
          newOffsetInMetadata.put(topicPartition, outOfRangeOffset);
          logger.warn("Resetting offset in metadata for "
                      + topicPartition
                      + " to earliest offset: "
                      + earliestAvailableOffset);
          sendResetRequestAndWait(CollectionUtils.mapKeys(
              newOffsetInMetadata,
              streamPartition -> StreamPartition.of(
                  streamPartition.topic(),
                  streamPartition.partition()
              )
          ), taskToolbox);
        } else {
          // With the offset not in range (earliestAvailableOffset, latestAvailableOffset), there is not much we can do
          // but wait for the available offsets in the partition to arrive in the range.
          logger.warn("Offset "
                      + outOfRangeOffset
                      + " is out of range of the available offsets for "
                      + topicPartition
                      + ". It is likely that a manual offset reset of the supervisor is needed");

          log.warn("Retrying in %dms", task.getPollRetryMs());
          pollRetryLock.lockInterruptibly();
          try {
            long nanos = TimeUnit.MILLISECONDS.toNanos(task.getPollRetryMs());
            while (nanos > 0L && !pauseRequested && !stopRequested.get()) {
              nanos = isAwaitingRetry.awaitNanos(nanos);
            }
          }
          finally {
            pollRetryLock.unlock();
          }
        }
```

Reply:
@gianm - could you explain the scenario in which the offset we are asking the consumer to seek to could possibly be higher than the latest available offset in Kafka? 
```
@Nonnull
  @Override
  protected List<OrderedPartitionableRecord<Integer, Long, KafkaRecordEntity>> getRecords(
      RecordSupplier<Integer, Long, KafkaRecordEntity> recordSupplier,
      TaskToolbox toolbox
  ) throws Exception
  {
    try {
      return recordSupplier.poll(task.getIOConfig().getPollTimeout());
    }
    catch (OffsetOutOfRangeException e) {
      //
      // Handles OffsetOutOfRangeException, which is thrown if the seeked-to
      // offset is not present in the topic-partition. This can happen if we're asking a task to read from data
      // that has not been written yet (which is totally legitimate). So let's wait for it to show up
      //
      log.warn("OffsetOutOfRangeException with message [%s]", e.getMessage());
      possiblyResetOffsetsOrWait(e.offsetOutOfRangePartitions(), recordSupplier, toolbox);
      return Collections.emptyList();
    }
  }
```

Reply:
For a partition that we've already committed some offset for, the starting offset given to a task for that partition is going to be the committed offset plus 1. So if no new messages have been written since the last commit, the starting offset will not exist yet.

Reply:
Thanks for the clarification, @gianm! 

@FrankChen021 - I would probably get rid of the log line I mentioned then and have the else block as it is (but moved as else condition for `if (outOfRangeOffset < earliestAvailableOffset)`

Reply:
@samarthjain I kept the code block as it was because original code reset the meta for all partition out of the partition loop and only reset for one time.